### PR TITLE
PXF extension with external table read from PXF demo profile

### PR DESCRIPTION
**Pull request by denalex**
This PR introduces a PXF extension that allows access to external data systems, such as Hadoop, using PXF server (that is currently part of Apache HAWQ (incubating)). Most of the logic here is a direct port from that project: https://github.com/apache/incubator-hawq/ with minimal changes.

This is a follow up on the previous PR (https://github.com/greenplum-db/gpdb/pull/2634). The logic currently implemented allows a user to define an external table with text columns using pxf:// protocol and PXF Demo profile (see README for more details). The setup requires PXF server to be running on all the machines where GPDB segments run. For the Demo profile, each GPDB segment is requesting one fragment and the PXF server returns 2 rows of demo (mock) data for each fragment.

To finally integrate this with Hadoop, we will need integration with real fragmenter and Hadoop namenode which will be accomplished in the next iterations.

For review: @d @pivotal-mike @schubert @danielgustafsson @hlinnaka @ascherbaum-pivotal @edespino @rvs @lavjain @sansanichfb @l-wang @Quikling @shivzone @kavinderd 

----------------------------------------------------------------
Signed-off-by: Lav Jain <ljain@pivotal.io>
Signed-off-by: Oleksandr Diachenko <odiachenko@pivotal.io>
Signed-off-by: Shivram Mani <smani@pivotal.io>
Signed-off-by: John Gaskin <jgaskin@pivotal.io>
Signed-off-by: Alexandra Wang <lewang@pivotal.io>

### CODE REVIEW

File: NOTICE [hunk 1]
```diff
@@ -2327,3 +2327,33 @@ documentation will at all times remain with copyright holders.
...
+ gpAux\extensions\pxf\src\pxfuriparser.c
+ gpAux\extensions\pxf\src\pxfuriparser.h
+ gpAux\extensions\pxf\src\pxfutils.c
+ gpAux\extensions\pxf\src\pxfutils.h
+ gpAux\extensions\pxf\test\pxfheaders_test.c
+ gpAux\extensions\pxf\test\pxfprotocol_test.c
+ gpAux\extensions\pxf\test\pxfuriparser_test.c
+ gpAux\extensions\pxf\test\pxfutils_test.c
+
+Apache HAWQ (incubating)
```
Comment (khannaekta) [inline NOTICE]: should we mention HAWQ or GPDB?

File: concourse/scripts/regression_tests_pxf.bash [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+	su gpadmin -c "bash /home/gpadmin/run_regression_test.sh $(pwd)"
+}
+
+function setup_gpadmin_user() {
+	./gpdb_src/concourse/scripts/setup_gpadmin_user.bash "$TARGET_OS"
+}
+
+function setup_singlecluster() {
+  pushd /home/gpadmin
+    wget https://s3-us-west-2.amazonaws.com/pivotal-singlecluster/singlecluster/singlecluster-HDP.tar.gz
```
Comment (khannaekta) [inline concourse/scripts/regression_tests_pxf.bash]: Do we need to parameterize the tarball? Will it not change often?

File: concourse/scripts/regression_tests_pxf.bash [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+		exit 1
+	fi
+
+	time configure
+	time install_gpdb
+	time setup_gpadmin_user
+	time make_cluster
+	time gen_env
+
+    time setup_singlecluster
```
Comment (khannaekta) [inline concourse/scripts/regression_tests_pxf.bash]: indent off.

File: gpAux/extensions/pxf/README.md [hunk 4]
```diff
@@ -0,0 +1,86 @@
+# The PXF extension client for GPDB
+
+At present, this extension is in development status, where only Demo PXF profile is working and no real data is yet accessible.
+
+## Table of Contents
+
+* Usage
```
Comment (khannaekta) [inline gpAux/extensions/pxf/README.md]: Indentation looks off.

File: gpAux/extensions/pxf/README.md [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+Additional instructions on building and starting a GPDB cluster can be
+found in the top-level [README.md](../../../README.md) ("_Build the
+database_" section).
+
+## Create and use PXF external table
+At this time, only PXF Demo profile is working:
+```
+# CREATE EXTERNAL TABLE pxf_read_test (a TEXT, b TEXT, c TEXT) \
+LOCATION ('pxf://localhost:51200/tmp/dummy1' \
+'?FRAGMENTER=org.apache.hawq.pxf.api.examples.DemoFragmenter' \
```
Comment (khannaekta) [inline gpAux/extensions/pxf/README.md]: in general, shouldn't this package be like org.apache.pxf...? without hawq? Somehow as we know HAWQ and GPDB as two separate db products, so the term hawq in gpdb pxf is a bit confusing.

File: NOTICE (hunk 1 above)
Comment (rvs) [inline NOTICE]: This is the code lifted from HAWQ (incubating)

File: gpAux/extensions/pxf/README.md (hunk 5 above)
Comment (rvs) [inline gpAux/extensions/pxf/README.md]: I'd leave it be for now. PXF is an external service that today happens to be a suproject of HAWQ. We may want to graduate it into its own TLP but that coversation is a few months off.

**Comment by hlinnaka** [comment]
In Greenplum, we rely on the ResourceOwner mechanism to close resources like open files, handles in external libraries like libcurl handles, etc. See commit df050f9bfd9aaf652d3851b3aa115b12e7d3b122, where we started doing that, for the changes that were done to the existing external protocols. In particular, PXF should track libcurl handles, like the code `url_curl.c` does. Or perhaps even better, PXF could use the existing code in `url_curl.c`. It probably needs some refactoring to make that possible, but seems best in the long run.

**Comment by kavinderd** [comment]
I like the approach of using an Extension to avoid having to change any executor code as was the case in HAWQ. 

Since PXF is completely new to GPDB, I think it would be a good time to rethink the use of HTTP headers for transmitting query information to the PXF webserver. We knew it was an inelegant approach in HAWQ since richer and richer query information would simply get written in plain text as an HTTP header. @sansanichfb has mentioned using Protobuf or some other serialization technology. I fear that if it is not done now it may never get done.

**Comment by rvs** [comment]
If you're re-thinking HTTP approach and have protobuf in mind, I would highly, highly recommend taking a look at gRPC https://grpc.io/

File: gpAux/extensions/pxf/Makefile [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+include $(PGXS)
+else
+	top_builddir = ../../..
+	include $(top_builddir)/src/Makefile.global
+	include $(top_srcdir)/contrib/contrib-global.mk
+endif
+
+CURL_DIR = /usr/local/opt/curl
+#override CPPFLAGS := -I$(CURL_DIR)/include $(CPPFLAGS)
+#override LDFLAGS := -L$(CURL_DIR)/lib $(LDFLAGS)
```
Comment (denalex) [inline gpAux/extensions/pxf/Makefile]: thanks, will do

File: concourse/scripts/regression_tests_pxf.bash (hunk 2 above)
Comment (denalex) [inline concourse/scripts/regression_tests_pxf.bash]: thanks, will change to use a resource in concourse.

**Comment by denalex** [comment]
I agree that HTTP headers are not the best way to send information and I think we need to research the alternatives for both HTTP transfer and the headers. However, any change will have a profound effect on existing code, so I would propose to consider these enhancements after the initial MVP, otherwise the feature ship date will move even further into the future.

**Comment by sansanichfb** [comment]
+1 to @denalex point. It definitely makes sense to improve the way we send information over HTTP, probably Protobuf. But I don't think it will have any benefits in shifting the paradigm from REST to RPC.

**Comment by denalex** [comment]
@hlinnaka -- thank you for the pointer to the commit and the logic to handle urls. We will review it, but I would also propose considering any changes to post-MVP, especially if we are to re-evaluate the transport protocol between GPDB and PXF.

**Comment by shivzone** [comment]
@kavinderd i do agree that the existing means of using http headers is not very manageable compared to any evolvable format like protobuf. Bringing in protobuf as an added dependency in gpdb might complicate the build process for starts. 
@rvs i would rather that the grpc interface (+protobuf) would definitely be a better choice over REST/JSON or equivalent. 
Would have to consider tradeoff between feature availability and time.

File: gpAux/extensions/pxf/src/pxfutils.c [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+
+/*
+ * TypeOidGetTypename
+ * Get the name of the type, given the OID
+ */
+char*
+TypeOidGetTypename(Oid typid)
+{
+
+	Assert(OidIsValid(typeid));
```
Comment (denalex) [inline gpAux/extensions/pxf/src/pxfutils.c]: thx, will fix.

File: gpAux/extensions/pxf/README.md [hunk 8]
```diff
@@ -0,0 +1,93 @@
+# The PXF extension client for GPDB
+
+At present, this extension is in development status, where only Demo PXF profile is working and no real data is yet accessible.
+
+## Table of Contents
+
+* Usage
+<<<<<<< HEAD
```
Comment (shivzone) [inline gpAux/extensions/pxf/README.md]: Please remove this diff

File: gpAux/extensions/pxf/README.md [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+=======
+
+## Usage
+
+### Enable PXF extension in GPDB build process.
+
+Configure GPDB to build the pxf extension by adding the `--with-pxf`
+configure option. This is required to setup the PXF build environment.
+
+### Build the PXF extension
```
Comment (sansanichfb) [inline gpAux/extensions/pxf/README.md]: We could also put more details about building pxf as a part of GPDB and as a standalone extension.

File: gpAux/extensions/pxf/src/pxfuriparser.h [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+#include "postgres.h"
+#include "fmgr.h"
+#include "utils/builtins.h"
+
+/*
+ * Path constants for accessing PXF.
+ * All PXF's resources are under /PXF_SERVICE_PREFIX/PXF_VERSION/...
+ */
+#define PXF_SERVICE_PREFIX "pxf"
+#define PXF_VERSION "v15" /* PXF version */
```
Comment (sansanichfb) [inline gpAux/extensions/pxf/src/pxfuriparser.h]: That looks more like PXF's protocol version rather than an actual version of PXF.

**Comment by lisakowen** [comment]
protocols are defined at the system/cluster-level, right?  extensions are defined/enabled at the database level.  would/can the admin create the PXF extension in more than one database in a Greenplum cluster?  

i noticed that the PXF read/write/validate functions are all created with "CREATE OR REPLACE FUNCTION", so that would not be an issue.  wasn't sure about the CREATE TRUSTED PROTOCOL pxf ...

**Pull request merged**